File: backend/handlers/tasks.php
<?php

function getTasks(PDO $databaseConnection): void
{
    $userId = requireAuth();

    $stmt = $databaseConnection->prepare(
        'SELECT id, task_text, priority, deadline, is_completed, created_at
         FROM tasks
         WHERE user_id = ?
         ORDER BY id DESC'
    );
    $stmt->execute([$userId]);

    $tasks = array_map(
        static fn($task) => [
            ...$task,
            'id' => (int)$task['id'],
            'priority' => (int) $task['priority'],
            'is_completed' => (bool)$task['is_completed'],
        ],
        $stmt->fetchAll()
    );

    echo json_encode($tasks);
}

function createTask(PDO $databaseConnection): void
{
    $userId = requireAuth();
    $data = json_decode(file_get_contents('php://input'), true) ?? [];

    $text = trim($data['task_text'] ?? '');
    $priority = (int) ($data['priority'] ?? 2);
    $deadline = ($data['deadline'] ?? '') ?: null;

    if (!$text) {
        http_response_code(422);
        echo json_encode(['error' => 'The task text field is required']);
        return;
    }

    if (!in_array($priority, [1, 2, 3], true)) {
        $priority = 2;
    }

    $stmt = $databaseConnection->prepare(
        'INSERT INTO tasks (user_id, task_text, priority, deadline) VALUES (?, ?, ?, ?)'
    );
    $stmt->execute([$userId, $text, $priority, $deadline]);

    http_response_code(201);
    echo json_encode([
        'id'           => (int)$databaseConnection->lastInsertId(),
        'task_text'    => $text,
        'priority'     => $priority,
        'deadline'     => $deadline,
        'is_completed' => false,
        'created_at'   => date('Y-m-d H:i:s'),
    ]);
}

function updateTask(PDO $databaseConnection, ?int $id): void
{
    $userId = requireAuth();

    if (!$id) {
        http_response_code(400);
        echo json_encode(['error' => 'Task ID required']);
        return;
    }

    $data = json_decode(file_get_contents('php://input'), true) ?? [];
    $fields = [];
    $values = [];

    if (array_key_exists('task_text', $data)) {
        $text = trim($data['task_text']);
        if (!$text) {
            http_response_code(422);
            echo json_encode(['error' => 'The task text field is required']);
            return;
        }
        $fields[] = 'task_text = ?';
        $values[] = $text;
    }

    if (array_key_exists('is_completed', $data)) {
        $fields[] = 'is_completed = ?';
        $values[] = (int)$data['is_completed'];
    }

    if (array_key_exists('priority', $data)) {
        $priority = (int) $data['priority'];
        $fields[] = 'priority = ?';
        $values[] = in_array($priority, [1, 2, 3], true) ? $priority : 2;
    }

    if (array_key_exists('deadline', $data)) {
        $fields[] = 'deadline = ?';
        $values[] = ($data['deadline'] ?: null);
    }

    if (!$fields) {
        http_response_code(422);
        echo json_encode(['error' => 'No fields to update']);
        return;
    }

    $values[] = $id;
    $values[] = $userId;

    $stmt = $databaseConnection->prepare(
        'UPDATE tasks SET ' . implode(', ', $fields) . ' WHERE id = ? AND user_id = ?'
    );
    $stmt->execute($values);

    echo json_encode(['updated' => $stmt->rowCount() > 0]);
}

function deleteTask(PDO $databaseConnection, ?int $id): void
{
    $userId = requireAuth();

    if (!$id) {
        http_response_code(400);
        echo json_encode(['error' => 'Task ID required']);
        return;
    }

    $stmt = $databaseConnection->prepare('DELETE FROM tasks WHERE id = ? AND user_id = ?');
    $stmt->execute([$id, $userId]);

    echo json_encode(['deleted' => $stmt->rowCount() > 0]);
}

File: backend/internal/handlers/tasks.go
package handlers

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"github.com/Andrii-K-17/just-tasks/internal/middleware"
	"github.com/Andrii-K-17/just-tasks/internal/response"
	"github.com/Andrii-K-17/just-tasks/internal/services"
	"github.com/go-chi/chi/v5"
)

// TaskHandler manages task-related HTTP endpoints.
type TaskHandler struct {
	svc *services.TaskService
}

// NewTaskHandler initializes and returns a new TaskHandler.
func NewTaskHandler(svc *services.TaskService) *TaskHandler {
	return &TaskHandler{svc: svc}
}

// GetTasks retrieves all owned and shared tasks for the authenticated user.
func (h *TaskHandler) GetTasks(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	tasks, err := h.svc.GetAll(userID)
	if err != nil {
		response.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	response.JSON(w, http.StatusOK, tasks)
}

// createTaskRequest represents the payload for creating a new task.
type createTaskRequest struct {
	TaskText   string  `json:"task_text"`
	Priority   int     `json:"priority"`
	Deadline   *string `json:"deadline"`
	CategoryID *int    `json:"category_id"`
}

// CreateTask inserts a new task into the database and returns it.
func (h *TaskHandler) CreateTask(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	var req createTaskRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		response.Error(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if strings.TrimSpace(req.TaskText) == "" {
		response.Error(w, http.StatusUnprocessableEntity, "the task text field is required")
		return
	}

	var deadline *string
	if req.Deadline != nil && *req.Deadline != "" {
		deadline = req.Deadline
	}

	task, err := h.svc.Create(userID, services.CreatePayload{
		TaskText:   req.TaskText,
		Priority:   req.Priority,
		Deadline:   deadline,
		CategoryID: req.CategoryID,
	})
	if err != nil {
		response.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	response.JSON(w, http.StatusCreated, task)
}

// UpdateTask modifies fields of an existing task belonging to the authenticated user.
func (h *TaskHandler) UpdateTask(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	id, err := strconv.Atoi(chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, http.StatusBadRequest, "invalid task id")
		return
	}

	var body map[string]any
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		response.Error(w, http.StatusBadRequest, "invalid request body")
		return
	}

	p := services.UpdatePayload{}

	if v, ok := body["task_text"]; ok {
		text := strings.TrimSpace(fmt.Sprintf("%v", v))
		if text == "" {
			response.Error(w, http.StatusUnprocessableEntity, "the task text field is required")
			return
		}
		p.TaskText = &text
	}
	if v, ok := body["is_completed"]; ok {
		b := v.(bool)
		p.IsCompleted = &b
	}
	if v, ok := body["priority"]; ok {
		pri := int(v.(float64))
		p.Priority = &pri
	}
	if v, ok := body["deadline"]; ok {
		if v == nil || v == "" {
			p.DeadlineNull = true
		} else {
			s := v.(string)
			p.Deadline = &s
		}
	}
	if v, ok := body["category_id"]; ok {
		if v == nil {
			p.CategoryNull = true
		} else {
			id := int(v.(float64))
			p.CategoryID = &id
		}
	}

	if err := h.svc.Update(userID, id, p); err != nil {
		switch {
		case errors.Is(err, services.ErrTaskNotFound):
			response.Error(w, http.StatusNotFound, "task not found")
		case errors.Is(err, services.ErrForbidden):
			response.Error(w, http.StatusForbidden, "forbidden")
		default:
			response.Error(w, http.StatusInternalServerError, "internal error")
		}
		return
	}

	response.JSON(w, http.StatusOK, map[string]bool{"updated": true})
}

// DeleteTask removes a task by its ID for the authenticated user.
func (h *TaskHandler) DeleteTask(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	id, err := strconv.Atoi(chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, http.StatusBadRequest, "invalid task id")
		return
	}

	if err := h.svc.Delete(userID, id); err != nil {
		if errors.Is(err, services.ErrForbidden) {
			response.Error(w, http.StatusForbidden, "forbidden or task not found")
			return
		}
		response.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	response.JSON(w, http.StatusOK, map[string]bool{"deleted": true})
}

// reorderRequest represents the payload for reordering tasks.
type reorderRequest struct {
	IDs []int `json:"ids"`
}

// ReorderTasks updates the position of tasks based on the provided order.
func (h *TaskHandler) ReorderTasks(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	var req reorderRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || len(req.IDs) == 0 {
		response.Error(w, http.StatusUnprocessableEntity, "ids array required")
		return
	}

	if err := h.svc.Reorder(userID, req.IDs); err != nil {
		response.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	response.JSON(w, http.StatusOK, map[string]bool{"ok": true})
}

// AddCollaborator adds a user to a task by their username.
func (h *TaskHandler) AddCollaborator(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	taskID, err := strconv.Atoi(chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, http.StatusBadRequest, "invalid task id")
		return
	}

	var req struct {
		Username string `json:"username"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Username == "" {
		response.Error(w, http.StatusBadRequest, "username required")
		return
	}

	added, err := h.svc.AddCollaborator(userID, taskID, req.Username)
	if err != nil {
		switch {
		case errors.Is(err, services.ErrForbidden):
			response.Error(w, http.StatusForbidden, "only owner can add collaborators")
		case errors.Is(err, services.ErrTaskNotFound):
			response.Error(w, http.StatusNotFound, "user not found")
		default:
			response.Error(w, http.StatusBadRequest, err.Error())
		}
		return
	}

	response.JSON(w, http.StatusOK, map[string]bool{"added": added})
}

// RemoveCollaborator removes a user from a task.
func (h *TaskHandler) RemoveCollaborator(w http.ResponseWriter, r *http.Request) {
	userID := middleware.GetUserID(r.Context())

	taskID, err := strconv.Atoi(chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, http.StatusBadRequest, "invalid task id")
		return
	}

	collabID, err := strconv.Atoi(chi.URLParam(r, "collabId"))
	if err != nil {
		response.Error(w, http.StatusBadRequest, "invalid collaborator id")
		return
	}

	removed, err := h.svc.RemoveCollaborator(userID, taskID, collabID)
	if err != nil {
		if errors.Is(err, services.ErrForbidden) {
			response.Error(w, http.StatusForbidden, "only owner can remove collaborators")
			return
		}
		response.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	response.JSON(w, http.StatusOK, map[string]bool{"removed": removed})
}
